### DISCUSSION TITLE
关于数组指针

### DISCUSSION THREAD

**Discussion by zhengzhch** [arcsysu/SYsU-lang#50] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/56723398/169704107-3f32d6d2-1b77-4641-b2b1-7208b6b859d5.png)
![image](https://user-images.githubusercontent.com/56723398/169704168-8280dc65-e706-4838-9f14-e7aadc1a4ce7.png)
对于上面这个测例，传入了一个数组指针，在用GEP寻址时，按照对二维数组的方法来会报错如下
![image](https://user-images.githubusercontent.com/56723398/169704209-0707df71-4d38-443e-8207-72c280181162.png)
推测可能是最里层解开后，变成了a[5]，不知道有什么方法能解决这个问题

**Comment by Yorkking** [comment, marked as answer]
ArrayType 使用 gep 访问时，需要额外的 0 索引，即调用 :
```C++
 Builder.CreateInBoundsGEP(
            baseArr,
            {llvm::ConstantInt::get(llvm::Type::getInt32Ty(TheContext), 0),
             Idx},
            "arrayIdx");
```

但是指针类型，用 gep 访问时，不需要额外的 0 索引，即:
```C++
Builder.CreateInBoundsGEP(basePtr, idx, "arrayIdx");
```

**Reply by zhengzhch** [reply]
那请问有API能判断一个Value*是否是个指针吗？

**Reply by zhengzhch** [reply]
哦哦，已经找到了，谢谢

**Reply by zhengzhch** [reply]
![image](https://user-images.githubusercontent.com/56723398/169731078-a9e3cf40-ed43-44b8-9b4a-a285f721446c.png)
但是对于上面这种情况，获得的也是指针。可是实际上却需要额外的0索引，不知道有没有什么办法把这种数组和指针区分开来
![X4DFO{0ORC0W0QNS5R)8U](https://user-images.githubusercontent.com/56723398/169731111-b2d07f2a-3ca5-42ef-bf0a-cf025abfa0a8.png)

**Reply by Yorkking** [reply]
下面同学说的有道理

**Comment by Thirtory** [comment]
GEP的功能其实应该就是如其名Get Element Pointer。我的理解是：
对于一个指针，比如以i32\*为例的话，GEP 0 的功能是取其偏移0 （看成2步，Get Element 0, Get Pointer），所以最后结果应该和原来是相等的（同为i32* 当然这部分没有实操过，属于推测）。

对于更复杂一点的情况，比如这里的a[0][0]其实是有多层的ArrayToPointerDecay（隐式转换 比如数组 [4 × [2 × i32]] 会被当成指针[2 × i32]\*）。llvm的做法像是按语法树一层层地递归向下取GEP。

具体分析来说，a是 [4 × [2 × i32]]\* ，它是一个指向[4 × [2 × i32]]数组的指针。第一行%arrayidx做了 GEP 0 0，前一个0取到了一个[4 × [2 × i32]]数组元素的指针（解指针为元素），以（如果有）隐式类型转换的视角来看数组的话 [4 × [2 × i32]] 类型也可以看做 [2 × i32]\* 类型，那么第二个0就是取 [2 × i32]\* 这个指针上0偏移的元素的指针 也就是一个 [2 × i32]\* 的指针了。
同理继续看%arrayidx2 ，首先上一步我们得到了%arrayidx是 [2 × i32]\* ，现在对其取GEP 0 0。第一维0仍然是解指针为元素，即得到一个[2 × i32]，然后以转换为指针的角度看就是i32\*，第二维0就是取这个 i32*指针上偏移为0的元素的指针（注意仍是i32\*）。

所以我的观点是，是否需要额外0索引不是取决于访问的数组类型还是指针类型（事实上都是\*（指针）吧，而且数组本身也很像一种指针=-= 我觉得数组类型需要额外0索引只是因为我们store的变量是一个指向数组的指针（比如这里的a其实可以看做一个[1 ×[4 × [2 × i32]]]的数组），需要GEP0来将其解开）。

BTW，%arrayidx和%arrayidx2可以合为一句GEP 0 0 0，即：第一维0将指向数组的指针解开成数组元素（当然它存的还是一个指向数组的指针），后面2维是下标索引。

**Reply by Yorkking** [reply]
确实应该这么理解。
